### PR TITLE
Prometheus exporter

### PR DESCRIPTION
**Pull request by sriniketh923**
Base code for Prometheus exporter

### CODE REVIEW

File: core/metrics/exporter.go [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+	}
+
+	for eventObject := range ch {
+		event := eventObject.Event
+		typeThing := eventObject.Type
+		var data interface{} = eventObject.Data
+		data = eventObject.Data
+		if event == "lock-core" {
+			log.Infof("To be done")
+			// coreStatusGauge.Set(1)
```
Comment (thomasmitchell) [inline core/metrics/exporter.go]: clean this up

File: core/metrics/exporter.go [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+		} else if event == "unlock-core" {
+			log.Infof("To be done")
+			// coreStatusGauge.Set(2)
```
Comment (thomasmitchell) [inline core/metrics/exporter.go]: clean this up

File: core/metrics/exporter.go [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+func (m *Metrics) RegisterExporter(tenantCount, agentsCount, targetsCount, storesCount, jobsCount, tasksCount, archivesCount float64) {
+	prometheus.MustRegister(tenantsGauge)
+	prometheus.MustRegister(agentsGauge)
+	prometheus.MustRegister(targetsGauge)
+	prometheus.MustRegister(storesGauge)
+	prometheus.MustRegister(jobsGauge)
+	prometheus.MustRegister(tasksGauge)
+	prometheus.MustRegister(archivesGauge)
+	prometheus.MustRegister(storageUsedBytesGauge)
+	/* To be done
```
Comment (thomasmitchell) [inline core/metrics/exporter.go]: Make TODO

File: core/metrics/exporter.go [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+
+	tenantsGauge.Set(tenantCount)
+	agentsGauge.Set(agentsCount)
+	targetsGauge.Set(targetsCount)
+	storesGauge.Set(storesCount)
+	jobsGauge.Set(jobsCount)
+	tasksGauge.Set(tasksCount)
+	archivesGauge.Set(archivesCount)
+	storageUsedBytesGauge.Set(0)
+	/* To be done
```
Comment (thomasmitchell) [inline core/metrics/exporter.go]: TODO

File: core/metrics/exporter.go [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+	prometheus.MustRegister(coreStatusGauge) */
+
+	tenantsGauge.Set(tenantCount)
+	agentsGauge.Set(agentsCount)
+	targetsGauge.Set(targetsCount)
+	storesGauge.Set(storesCount)
+	jobsGauge.Set(jobsCount)
+	tasksGauge.Set(tasksCount)
+	archivesGauge.Set(archivesCount)
+	storageUsedBytesGauge.Set(0)
```
Comment (thomasmitchell) [inline core/metrics/exporter.go]: This should probably not be initialized to zero every time.

File: core/main.go [hunk 6]
```diff
@@ -22,6 +23,7 @@ func (c Core) Main() {
 
 	/* we need a usable database first */
 	c.ConnectToDatabase()
+	c.InitalizePrometheus() //Initalize metirc values
```
Comment (thomasmitchell) [inline core/main.go]: Should be `InitializePrometheus`

File: core/main.go [hunk 7]
```diff
@@ -125,6 +128,46 @@ func (c *Core) ConnectToDatabase() {
 	log.Debugf("connected successfully to database!")
 }
 
+func (c *Core) InitalizePrometheus() {
```
Comment (thomasmitchell) [inline core/main.go]: Should be `InitializePrometheus`

File: core/main.go [hunk 8]
```diff
@@ -125,6 +128,46 @@ func (c *Core) ConnectToDatabase() {
...
+	tenants, err := c.db.GetAllTenants(nil)
+	if err != nil {
+		log.Debugf("No tenats found in database")
```
Comment (thomasmitchell) [inline core/main.go]: No `tenants` ...

File: core/metrics/exporter.go [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+func InitalizeMetrics() *Metrics {
+	metrics := &Metrics{}
+	return metrics
+}
+
+func (m *Metrics) Inform(mbus *bus.Bus) {
+	m.bus = mbus
+}
+
+func (m *Metrics) RegisterExporter(tenantCount, agentsCount, targetsCount, storesCount, jobsCount, tasksCount, archivesCount float64) {
```
Comment (thomasmitchell) [inline core/metrics/exporter.go]: Change input type to int instead of float. Cast within function.

File: core/metrics/exporter.go (hunk 5 above)
Comment (thomasmitchell) [reply core/metrics/exporter.go]: Probably need to add an extra parameter to the function to pass in this value.

File: core/metrics/exporter.go (hunk 9 above)
Comment (thomasmitchell) [reply core/metrics/exporter.go]: You should probably change all of the args here into a single struct - when you have this many args to a function, remembering the ordering becomes infeasible. Putting it into a struct makes the caller label the inputs, and also allows them to specify them in any order.

File: core/metrics/exporter.go [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+		})
+
+	storageUsedBytesGauge = prometheus.NewGauge(
+		prometheus.GaugeOpts{
+			Namespace: namespace,
+			Name:      storageBytesTotal,
+			Help:      "How much storage has been used, in bytes.",
+		})
+
+	/* To be implemented
```
Comment (thomasmitchell) [inline core/metrics/exporter.go]: Change to TODO

File: core/metrics/exporter.go [hunk 11]
```diff
@@ -0,0 +1,235 @@
...
+	namespace         = "shield"
+	tenantsTotal      = "tenants_total"
+	agentsTotal       = "agents_total"
+	targetsTotal      = "targets_total"
+	storesTotal       = "stores_total"
+	jobsTotal         = "jobs_total"
+	tasksTotal        = "tasks_total"
+	archivesTotal     = "archives_total"
+	storageBytesTotal = "storage_used_bytes"
+	/* To be implemented
```
Comment (thomasmitchell) [inline core/metrics/exporter.go]: Change to TODO

File: core/main.go [hunk 12]
```diff
@@ -125,6 +128,46 @@ func (c *Core) ConnectToDatabase() {
...
+	if err != nil {
+		log.Debugf("No tasks found in database")
+	}
+
+	archives, err := c.db.GetAllArchives(nil)
+	if err != nil {
+		log.Debugf("No archives found in database")
+	}
+
+	c.metrics = metrics.InitalizeMetrics()
```
Comment (thomasmitchell) [inline core/main.go]: Should be `InitializeMetrics`

File: core/metrics/exporter.go [hunk 13]
```diff
@@ -0,0 +1,235 @@
...
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"github.com/shieldproject/shield/core/bus"
+)
+
+type Metrics struct {
+	bus *bus.Bus
+}
+
+const (
+	namespace         = "shield"
```
Comment (thomasmitchell) [inline core/metrics/exporter.go]: namespace should be configurable, and therefore cannot be a const

File: core/metrics/exporter.go [hunk 14]
```diff
@@ -0,0 +1,235 @@
...
+	tasksTotal        = "tasks_total"
+	archivesTotal     = "archives_total"
+	storageBytesTotal = "storage_used_bytes"
+	/* To be implemented
+	targetHealthStatus = "target_health_status"
+	storeHealthStatus  = "storeHealthStatus"
+	coreStatus         = "core_status" */
+)
+
+var (
```
Comment (thomasmitchell) [inline core/metrics/exporter.go]: These should be members of the metrics struct, not singletons

File: core/metrics/exporter.go [hunk 15]
```diff
@@ -0,0 +1,235 @@
...
+	coreStatusGauge = prometheus.NewGauge(
+		prometheus.GaugeOpts{
+			Namespace: namespace,
+			Name:      coreStatus,
+			Help:      "The global initialized / locked status of the SHIELD core; (0=uninitialized, 1=locked, 2=unlocked)",
+		})
+	*/
+)
+
+func InitalizeMetrics() *Metrics {
```
Comment (thomasmitchell) [inline core/metrics/exporter.go]: Idiomatically, this should just be called `New`

Comment (thomasmitchell) [reply core/metrics/exporter.go]: so it would read `metrics.New()`

File: core/metrics/exporter.go (hunk 9 above)
Comment (thomasmitchell) [reply core/metrics/exporter.go]: This entire function should just be part of InitializeMetrics... which should have its name changed to `New`

File: core/main.go [hunk 16]
```diff
@@ -22,6 +23,7 @@ func (c Core) Main() {
 
 	/* we need a usable database first */
 	c.ConnectToDatabase()
+	c.InitializePrometheus() //Initialize metric values
```
Comment (thomasmitchell) [inline core/main.go]: The exported call from the package should be `metrics.New(...)`

File: core/metrics/exporter.go [hunk 17]
```diff
@@ -0,0 +1,222 @@
...
+import (
+	"net/http"
+
+	"github.com/jhunt/go-log"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"github.com/shieldproject/shield/core/bus"
+)
+
+type Config struct {
```
Comment (thomasmitchell) [inline core/metrics/exporter.go]: Put namespace in the config.

File: core/metrics/exporter.go [hunk 18]
```diff
@@ -0,0 +1,222 @@
...
+	tasksTotal        = "tasks_total"
+	archivesTotal     = "archives_total"
+	storageBytesTotal = "storage_used_bytes"
+	/* TODO
+	targetHealthStatus = "target_health_status"
+	storeHealthStatus  = "storeHealthStatus"
+	coreStatus         = "core_status" */
+)
+
+func New(namespace string, config Config) *Metrics {
```
Comment (thomasmitchell) [inline core/metrics/exporter.go]: move namespace into the Config.

File: core/main.go [hunk 19]
```diff
@@ -125,6 +128,61 @@ func (c *Core) ConnectToDatabase() {
...
+	}
+
+	storageBytesUsed, err := c.db.ArchiveStorageFootprint(&db.ArchiveFilter{
+		WithStatus: []string{"valid"},
+	})
+	if err != nil {
+		log.Debugf("Could not calculate bytes used by the storage")
+	}
+
+	c.metrics = metrics.New("", metrics.Config{
```
Comment (thomasmitchell) [inline core/main.go]: can we hook up the namespace to the core config?

File: core/metrics/exporter.go [hunk 20]
```diff
@@ -0,0 +1,222 @@
...
+		} else if event == "unlock-core" {
+			log.Infof("TODO")
+		} else if event == "create-object" {
+			m.CreateObjectCount(typeThing, data)
+		} else if event == "update-object" {
+			m.UpdateObjectCount(typeThing, data)
+		} else if event == "delete-object" {
+			m.DeleteObjectCount(typeThing)
+		} else {
+			log.Infof("Event not recognized yet")
```
Comment (thomasmitchell) [inline core/metrics/exporter.go]: ``Metrics ignoring event of type `%s'``, where `%s` is the `event` variable.
```suggestion
			log.Debugf("Metrics ignoring event of type `%s'", event)
```

File: core/metrics/exporter.go [hunk 21]
```diff
@@ -0,0 +1,222 @@
...
+		m.storesGauge.Inc()
+	case "job":
+		m.jobsGauge.Inc()
+	case "task":
+		m.tasksGauge.Inc()
+	case "archive":
+		m.archivesGauge.Inc()
+		m.storageUsedBytesGauge.Add(float64(interfaceData["size"].(int64)))
+	default:
+		log.Infof("Event type not recognized or not implemented yet.")
```
Comment (thomasmitchell) [inline core/metrics/exporter.go]: ``Metrics ignoring create event for object type `%s'`` where `%s` is the variable `typeThing`
```suggestion
		log.Debugf("Metrics ignoring create event for object type `%s'", typeThing)
```

Comment (thomasmitchell) [reply core/metrics/exporter.go]: Also, change to a debug level log.

File: core/metrics/exporter.go (hunk 20 above)
Comment (thomasmitchell) [reply core/metrics/exporter.go]: Change to a debug level log.

File: core/metrics/exporter.go [hunk 22]
```diff
@@ -0,0 +1,222 @@
...
+	case "agent":
+		m.agentsGauge.Dec()
+	case "target":
+		m.targetsGauge.Dec()
+	case "store":
+		m.storesGauge.Dec()
+	case "job":
+		m.jobsGauge.Dec()
+	default:
+		log.Infof("Event type not recognized or not implemented yet.")
```
Comment (thomasmitchell) [inline core/metrics/exporter.go]: Change to a debug level log.

Comment (thomasmitchell) [reply core/metrics/exporter.go]:
```suggestion
		log.Debugf("Metrics ignoring create event for object type `%s'", typeThing)
```

File: core/metrics/exporter.go [hunk 23]
```diff
@@ -0,0 +1,222 @@
...
+
+func (m *Metrics) UpdateObjectCount(typeThing string, data interface{}) {
+	interfaceData := data.(map[string]interface{})
+	switch typeThing {
+	case "archive":
+		if interfaceData["status"] == "manually purged" {
+			m.archivesGauge.Dec()
+			m.storageUsedBytesGauge.Sub(float64(interfaceData["size"].(int64)))
+		}
+	}
```
Comment (thomasmitchell) [inline core/metrics/exporter.go]: Add similar debug line here for unknown object type, if we're doing that.

File: core/core.go [hunk 24]
```diff
@@ -136,6 +139,7 @@ func init() {
 	DefaultConfig.DataDir = "/shield/data"
 	DefaultConfig.WebRoot = "/shield/ui"
 	DefaultConfig.PluginPathsEnv = "/shield/plugins"
+	DefaultConfig.PrometheusNamespace = ""
```
Comment (thomasmitchell) [inline core/core.go]: Make it `shield`

File: core/main.go [hunk 25]
```diff
@@ -125,6 +128,62 @@ func (c *Core) ConnectToDatabase() {
 	log.Debugf("connected successfully to database!")
 }
 
+func (c *Core) InitializePrometheus() {
+	tenants, err := c.db.GetAllTenants(nil)
+	if err != nil {
+		log.Debugf("No tenants found in database")
```
Comment (thomasmitchell) [inline core/main.go]: The error doesn't mean that there are no tenants. It means that there was a problem fetching the tenants.

Comment (thomasmitchell) [reply core/main.go]: Use `MaybeTerminate` to bail if this goes wrong.

Comment (thomasmitchell) [reply core/main.go]: Do this for all the instances in this function.

File: core/main.go [hunk 26]
```diff
@@ -274,6 +333,7 @@ func (c *Core) Bind() {
 	http.Handle("/v1/", c.v1API())
 	http.Handle("/v2/", c.v2API())
 	http.Handle("/auth/", c.authAPI())
+	http.Handle("/metrics/", c.metrics.ServeExporter()) //serve prometheus metircs at this endpoint
```
Comment (thomasmitchell) [inline core/main.go]: metrics is spelled wrong in the comment :)

File: core/metrics/exporter.go [hunk 27]
```diff
@@ -0,0 +1,226 @@
...
+	/* TODO
+	targetHealthStatus = "target_health_status"
+	storeHealthStatus  = "storeHealthStatus"
+	coreStatus         = "core_status" */
+)
+
+func New(config Config) *Metrics {
+	metrics := &Metrics{}
+	namespace := config.Namespace
+	if namespace == "" {
```
Comment (thomasmitchell) [inline core/metrics/exporter.go]: Move the defaulting to the config reading code, not here.

File: core/core.go [hunk 28]
```diff
@@ -41,11 +43,12 @@ type Core struct {
...
-	DataDir        string   `yaml:"data-dir"       env:"SHIELD_DATA_DIR"`
-	WebRoot        string   `yaml:"web-root"       env:"SHIELD_WEB_ROOT"`
-	PluginPaths    []string `yaml:"plugin_paths"`
-	PluginPathsEnv string   `yaml:"-"              env:"SHIELD_PLUGIN_PATHS"`
+	Debug               bool     `yaml:"debug"          env:"SHIELD_DEBUG"`
+	DataDir             string   `yaml:"data-dir"       env:"SHIELD_DATA_DIR"`
+	WebRoot             string   `yaml:"web-root"       env:"SHIELD_WEB_ROOT"`
+	PluginPaths         []string `yaml:"plugin_paths"`
+	PluginPathsEnv      string   `yaml:"-"              env:"SHIELD_PLUGIN_PATHS"`
+	PrometheusNamespace string
```
Comment (thomasmitchell) [inline core/core.go]:
```suggestion
	Prometheus struct {
	    Namespace string `yaml:"namespace"`
	} `yaml:"prometheus"`
```

File: core/core.go [hunk 29]
```diff
@@ -136,6 +139,7 @@ func init() {
 	DefaultConfig.DataDir = "/shield/data"
 	DefaultConfig.WebRoot = "/shield/ui"
 	DefaultConfig.PluginPathsEnv = "/shield/plugins"
+	DefaultConfig.PrometheusNamespace = "shield"
```
Comment (thomasmitchell) [inline core/core.go]: This will need to be edited slightly after the config name change.

File: core/main.go [hunk 30]
```diff
@@ -125,6 +128,72 @@ func (c *Core) ConnectToDatabase() {
 	log.Debugf("connected successfully to database!")
 }
 
+func (c *Core) InitializePrometheus() error {
+	tenants, err := c.db.GetAllTenants(nil)
+	if err != nil {
```
Comment (thomasmitchell) [inline core/main.go]: MaybeTerminate checks the error for you, so you don't need `if err != nil`

File: core/main.go [hunk 31]
```diff
@@ -125,6 +128,72 @@ func (c *Core) ConnectToDatabase() {
...
+		log.Errorf("Error initializing prometheus exporter: %s", err)
```
Comment (thomasmitchell) [inline core/main.go]: No need to `log.Errorf` if we're going to bail, because `Terminate` will `Alertf`

File: core/main.go (hunk 30 above)
Comment (thomasmitchell) [reply core/main.go]: Make these changes for all instances.

File: core/main.go (hunk 31 above)
Comment (thomasmitchell) [reply core/main.go]: Make these changes for all instances.

File: core/metrics/exporter.go [hunk 32]
```diff
@@ -0,0 +1,223 @@
...
+		})
+
+	metrics.storageUsedBytesGauge = prometheus.NewGauge(
+		prometheus.GaugeOpts{
+			Namespace: namespace,
+			Name:      storageBytesTotal,
+			Help:      "How much storage has been used, in bytes.",
+		})
+
+	prometheus.MustRegister(metrics.tenantsGauge)
```
Comment (thomasmitchell) [inline core/metrics/exporter.go]: MustRegister takes a variadic list. You can provide all of this registration as one call to `MustRegister`.

File: core/metrics/exporter.go [hunk 33]
```diff
@@ -0,0 +1,218 @@
...
+func (m *Metrics) UpdateObjectCount(typeThing string, data interface{}) {
+	interfaceData := data.(map[string]interface{})
+	switch typeThing {
+	case "archive":
+		if interfaceData["status"] == "manually purged" {
+			m.archivesGauge.Dec()
+			m.storageUsedBytesGauge.Sub(float64(interfaceData["size"].(int64)))
+		}
+	default:
+		log.Debugf("Metrics ignoring create event for object type `%s'", typeThing)
```
Comment (thomasmitchell) [inline core/metrics/exporter.go]: should be `update event`, not `create event`

File: core/metrics/exporter.go [hunk 34]
```diff
@@ -0,0 +1,218 @@
...
+	case "agent":
+		m.agentsGauge.Dec()
+	case "target":
+		m.targetsGauge.Dec()
+	case "store":
+		m.storesGauge.Dec()
+	case "job":
+		m.jobsGauge.Dec()
+	default:
+		log.Debugf("Metrics ignoring create event for object type `%s'", typeThing)
```
Comment (thomasmitchell) [inline core/metrics/exporter.go]: should be `delete event` not `create event`

File: core/core.go [hunk 35]
```diff
@@ -122,6 +124,10 @@ type Config struct {
 		Backlog  int `yaml:"backlog"   env:"SHIELD_MBUS_BACKLOG"`
 	} `yaml:"mbus"`
 
+	Prometheus struct {
+		Namespace string `yaml:"namespace`
+	}
```
Comment (thomasmitchell) [inline core/core.go]:
```suggestion
	} `yaml:"prometheus"`
```

File: core/core.go [hunk 36]
```diff
@@ -122,6 +124,10 @@ type Config struct {
 		Backlog  int `yaml:"backlog"   env:"SHIELD_MBUS_BACKLOG"`
 	} `yaml:"mbus"`
 
+	Prometheus struct {
+		Namespace string `yaml:"namespace`
```
Comment (thomasmitchell) [inline core/core.go]:
```suggestion
		Namespace string `yaml:"namespace"`
```

File: core/metrics/exporter.go [hunk 37]
```diff
@@ -0,0 +1,218 @@
...
+		log.Debugf("Metrics ignoring delete event for object type `%s'", typeThing)
+	}
+}
+
+/* TODO
+func (m *Metrics) UpdateCoreStatus(value float64) {
+	coreStatusGauge.Set(value)
+} */
+
+func (m *Metrics) RegisterBusEvents(queues ...string) {
```
Comment (jhunt) [inline core/metrics/exporter.go]: The name of this function is a bit misleading; Watch() would be better.

File: core/main.go [hunk 38]
```diff
@@ -36,6 +38,7 @@ func (c Core) Main() {
 	c.ConnectToVault()
 	c.Bind()
 	c.StartScheduler()
+	go c.metrics.RegisterBusEvents("*")
```
Comment (jhunt) [inline core/main.go]: I'm going to go out on a limb here and say that this doesn't compile any more?

Comment (sriniketh923) [reply core/main.go]: That was definitely a mistake on my part. It has been fixed now.

**Pull request merged**